cpp/src/parquet/column_writer.cc

```diff
@@ -410,25 +410,28 @@ class BufferedPageWriter : public PageWriter {
                      MemoryPool* pool = arrow::default_memory_pool(),
                      std::shared_ptr<Encryptor> meta_encryptor = nullptr,
                      std::shared_ptr<Encryptor> data_encryptor = nullptr)
-      : final_sink_(sink), metadata_(metadata) {
+      : final_sink_(sink), metadata_(metadata), has_dictionary_pages(false) {
     in_memory_sink_ = CreateOutputStream(pool);
     pager_ = std::unique_ptr<SerializedPageWriter>(new SerializedPageWriter(
         in_memory_sink_, codec, compression_level, metadata, row_group_ordinal,
         current_column_ordinal, pool, meta_encryptor, data_encryptor));
   }
 
   int64_t WriteDictionaryPage(const DictionaryPage& page) override {
+    has_dictionary_pages = true;
     return pager_->WriteDictionaryPage(page);
   }
 
   void Close(bool has_dictionary, bool fallback) override {
     // index_page_offset = -1 since they are not supported
     int64_t final_position = -1;
     PARQUET_THROW_NOT_OK(final_sink_->Tell(&final_position));
-    metadata_->Finish(
-        pager_->num_values(), pager_->dictionary_page_offset() + final_position, -1,
-        pager_->data_page_offset() + final_position, pager_->total_compressed_size(),
-        pager_->total_uncompressed_size(), has_dictionary, fallback);
+    auto dictionary_page_offset =
+        has_dictionary_pages ? pager_->dictionary_page_offset() + final_position : 0;
```

Comment:
It's a little bit janky that 0 is used as a "do not set the field" sentinel to the `Finish` method. Probably should add a comment here at least about this